redis/cluster.py

```diff
@@ -0,0 +1,1884 @@
+                    'ERROR sending "config get cluster-require-full-coverage"'
+                    ' command to redis server: {0}, {1}'.format(node.name, e)
+                )
+
+        # at least one node should have cluster-require-full-coverage yes
+        return any(node_require_full_coverage(node)
+                   for node in cluster_nodes.values())
+
+    def check_slots_coverage(self, slots_cache):
+        # Validate if all slots are covered or if we should try next
+        # startup node
+        for i in range(0, REDIS_CLUSTER_HASH_SLOTS):
+            if i not in slots_cache:
+                return False
+        return True
+
+    def create_redis_connections(self, nodes):
+        """
+        This function will create a redis connection to all nodes in :nodes:
+        """
+        for node in nodes:
+            if node.redis_connection is None:
+                node.redis_connection = self.create_redis_node(
+                    host=node.host,
+                    port=node.port,
+                    **self.connection_kwargs,
+                )
+
+    def create_redis_node(self, host, port, **kwargs):
+        if self.from_url:
+            # Create a redis node with a costumed connection pool
+            kwargs.update({"host": host})
+            kwargs.update({"port": port})
+            connection_pool = ConnectionPool(**kwargs)
+            r = Redis(
+                connection_pool=connection_pool
+            )
+        else:
+            r = Redis(
+                host=host,
+                port=port,
+                **kwargs
+            )
+        return r
+
+    def initialize(self):
+        """
+        Initializes the nodes cache, slots cache and redis connections.
+        :startup_nodes:
+            Responsible for discovering other nodes in the cluster
+        """
+        self.reset()
+        tmp_nodes_cache = {}
+        tmp_slots = {}
+        disagreements = []
+        startup_nodes_reachable = False
+        kwargs = self.connection_kwargs
+        for startup_node in self.startup_nodes.values():
+            try:
+                if startup_node.redis_connection:
+                    r = startup_node.redis_connection
+                else:
+                    # Create a new Redis connection and let Redis decode the
+                    # responses so we won't need to handle that
+                    copy_kwargs = copy.deepcopy(kwargs)
+                    copy_kwargs.update({"decode_responses": True})
+                    copy_kwargs.update({"encoding": "utf-8"})
+                    r = self.create_redis_node(
+                        startup_node.host, startup_node.port, **copy_kwargs)
+                    self.startup_nodes[startup_node.name].redis_connection = r
+                cluster_slots = r.execute_command("CLUSTER SLOTS")
+                startup_nodes_reachable = True
+            except (ConnectionError, TimeoutError):
+                continue
```

Comment:
Good example of where I'd like the logging restored - or at least added. A user has no idea otherwise.